### PR TITLE
Report error in status if more than one DataScienceCluster installed

### PR DESCRIPTION
**Pull request by ykaliuta**
Report error in status if more than one DataScienceCluster installed

V2
- update `phase` to `phaseError` as well
- add reporting for DSCInitialization

## How Has This Been Tested?

- Installed 2 DSCs in the same namespace with different name
- Checked `oc describe datasciencecluster.datasciencecluster.opendatahub.io/<NAME>` for both clusters

Both clusters report like 

```
Status:
  Conditions:
    Last Heartbeat Time:   2023-11-01T08:33:42Z
    Last Transition Time:  2023-11-01T08:32:20Z
    Message:               only one instance of DataScienceCluster object is allowed. Update existing instance on namespace  and name ykaliuta.dsc2
    Reason:                DuplicateDataScienceCluster
    Status:                False
    Type:                  ReconcileComplete
    Last Heartbeat Time:   2023-11-01T08:33:42Z
    Last Transition Time:  2023-11-01T08:32:20Z
    Message:               only one instance of DataScienceCluster object is allowed. Update existing instance on namespace  and name ykaliuta.dsc2
    Reason:                DuplicateDataScienceCluster
    Status:                False
    Type:                  Available
    Last Heartbeat Time:   2023-11-01T08:33:42Z
    Last Transition Time:  2023-11-01T08:31:39Z
    Message:               only one instance of DataScienceCluster object is allowed. Update existing instance on namespace  and name ykaliuta.dsc2
    Reason:                DuplicateDataScienceCluster
    Status:                False
    Type:                  Progressing
    Last Heartbeat Time:   2023-11-01T08:33:42Z
    Last Transition Time:  2023-11-01T08:32:20Z
    Message:               only one instance of DataScienceCluster object is allowed. Update existing instance on namespace  and name ykaliuta.dsc2
    Reason:                DuplicateDataScienceCluster
    Status:                True
    Type:                  Degraded
    Last Heartbeat Time:   2023-11-01T08:33:42Z
    Last Transition Time:  2023-11-01T08:32:20Z
    Message:               only one instance of DataScienceCluster object is allowed. Update existing instance on namespace  and name ykaliuta.dsc2
    Reason:                DuplicateDataScienceCluster
    Status:                Unknown
    Type:                  Upgradeable
    Last Heartbeat Time:   2023-11-01T08:32:12Z
    Last Transition Time:  2023-11-01T08:32:12Z
    Message:               Component reconciled successfully
    Reason:                ReconcileCompleted
    Status:                True
    Type:                  workbenchesReady
```

For DSCInitialization checked both DataScinceCluster:

```

  status:
    conditions:
    - lastHeartbeatTime: "2023-11-01T20:26:35Z"
      lastTransitionTime: "2023-11-01T20:26:14Z"
      message: only one instance of DSCInitialization object is allowed
      reason: DuplicateDSCInitialization
      status: "False"
      type: ReconcileComplete
    - lastHeartbeatTime: "2023-11-01T20:26:35Z"
      lastTransitionTime: "2023-11-01T20:26:14Z"
      message: only one instance of DSCInitialization object is allowed
      reason: DuplicateDSCInitialization
      status: "False"
      type: Available
    - lastHeartbeatTime: "2023-11-01T20:26:35Z"
      lastTransitionTime: "2023-11-01T20:26:14Z"
      message: only one instance of DSCInitialization object is allowed
      reason: DuplicateDSCInitialization
      status: "False"
      type: Progressing
    - lastHeartbeatTime: "2023-11-01T20:26:35Z"
      lastTransitionTime: "2023-11-01T20:26:14Z"
      message: only one instance of DSCInitialization object is allowed
      reason: DuplicateDSCInitialization
      status: "True"
      type: Degraded
    - lastHeartbeatTime: "2023-11-01T20:26:35Z"
      lastTransitionTime: "2023-11-01T20:26:14Z"
      message: only one instance of DSCInitialization object is allowed
      reason: DuplicateDSCInitialization
      status: Unknown
      type: Upgradeable
    phase: Error
```

and DSCIinitialization:

```status:
  conditions:
  - lastHeartbeatTime: "2023-11-01T20:25:08Z"
    lastTransitionTime: "2023-11-01T20:23:46Z"
    message: only one instance of DSCInitialization object is allowed. Update existing
      instance name new
    reason: DuplicateDSCInitialization
    status: "False"
    type: ReconcileComplete
  - lastHeartbeatTime: "2023-11-01T20:25:08Z"
    lastTransitionTime: "2023-11-01T20:23:46Z"
    message: only one instance of DSCInitialization object is allowed. Update existing
      instance name new
    reason: DuplicateDSCInitialization
    status: "False"
    type: Available
  - lastHeartbeatTime: "2023-11-01T20:25:08Z"
    lastTransitionTime: "2023-11-01T20:19:40Z"
    message: only one instance of DSCInitialization object is allowed. Update existing
      instance name new
    reason: DuplicateDSCInitialization
    status: "False"
    type: Progressing
  - lastHeartbeatTime: "2023-11-01T20:25:08Z"
    lastTransitionTime: "2023-11-01T20:23:46Z"
    message: only one instance of DSCInitialization object is allowed. Update existing
      instance name new
    reason: DuplicateDSCInitialization
    status: "True"
    type: Degraded
  - lastHeartbeatTime: "2023-11-01T20:25:08Z"
    lastTransitionTime: "2023-11-01T20:23:46Z"
    message: only one instance of DSCInitialization object is allowed. Update existing
      instance name new
    reason: DuplicateDSCInitialization
    status: Unknown
    type: Upgradeable
  phase: Error
```

## Merge criteria:

- [ ] The commits are squashed in a cohesive manner and have meaningful messages.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

**Comment by ykaliuta** [comment]
Actually, should be revisited after issue #645 integration.

**Comment by zdtsw** [comment]
ref: https://github.com/opendatahub-io/opendatahub-operator/issues/644

**Comment by zdtsw** [comment]
do we want to add the same status update for DSCI as well?

**Comment by ykaliuta** [comment]
> do we want to add the same status update for DSCI as well?

Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.

Well, if there are no  objections I'll add it as well. To both DSCI and DSC `Reconcile`.

**Comment by zdtsw** [comment]
> > do we want to add the same status update for DSCI as well?
> 
> Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.
> 
> Well, if there are no objections I'll add it as well. To both DSCI and DSC `Reconcile`.

ok, lets do like this way. 
we can get this out first and create a new issue for the DSCI, then have another PR for that issue, how you feel?

**Comment by ykaliuta** [comment]
> > > do we want to add the same status update for DSCI as well?
> > 
> > 
> > Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.
> > Well, if there are no objections I'll add it as well. To both DSCI and DSC `Reconcile`.
> 
> ok, lets do like this way. we can get this out first and create a new issue for the DSCI, then have another PR for that issue, how you feel?

I'm ok with several patches in one PR.

Just thought a bit about that. May be there should exist some sort of Webhook to prevent such configurations at the first place?

**Comment by zdtsw** [comment]
> > > > do we want to add the same status update for DSCI as well?
> > > 
> > > 
> > > Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.
> > > Well, if there are no objections I'll add it as well. To both DSCI and DSC `Reconcile`.
> > 
> > 
> > ok, lets do like this way. we can get this out first and create a new issue for the DSCI, then have another PR for that issue, how you feel?
> 
> I'm ok with several patches in one PR.
> 
> Just thought a bit about that. May be there should exist some sort of Webhook to prevent such configurations at the first place?

i do not want to enforce git webhook (if this is the webhook you meant) or it will be a new commit everytime and difficult for reviewer, we do have squash-merge for PR so the committer (or reviewer) is responsibile to get a clean history for the PR before merge

**Comment by ykaliuta** [comment]
> > > > > do we want to add the same status update for DSCI as well?
> > > > 
> > > > 
> > > > Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.
> > > > Well, if there are no objections I'll add it as well. To both DSCI and DSC `Reconcile`.
> > > 
> > > 
> > > ok, lets do like this way. we can get this out first and create a new issue for the DSCI, then have another PR for that issue, how you feel?
> > 
> > 
> > I'm ok with several patches in one PR.
> > Just thought a bit about that. May be there should exist some sort of Webhook to prevent such configurations at the first place?
> 
> i do not want to enforce git webhook (if this is the webhook you meant) or it will be a new commit everytime and difficult for reviewer, we do have squash-merge for PR so the committer (or reviewer) is responsibile to get a clean history for the PR before merge

No, I meant kubernetes validating webhook, like https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation

**Comment by zdtsw** [comment]
> > > > > > do we want to add the same status update for DSCI as well?
> > > > > 
> > > > > 
> > > > > Good question. Probably it will not do any harm. Just IIUC it is more complicated for the user to encounter. Like install the operator twice. When deploying twice just the DataScienceCluster is much easier.
> > > > > Well, if there are no objections I'll add it as well. To both DSCI and DSC `Reconcile`.
> > > > 
> > > > 
> > > > ok, lets do like this way. we can get this out first and create a new issue for the DSCI, then have another PR for that issue, how you feel?
> > > 
> > > 
> > > I'm ok with several patches in one PR.
> > > Just thought a bit about that. May be there should exist some sort of Webhook to prevent such configurations at the first place?
> > 
> > 
> > i do not want to enforce git webhook (if this is the webhook you meant) or it will be a new commit everytime and difficult for reviewer, we do have squash-merge for PR so the committer (or reviewer) is responsibile to get a clean history for the PR before merge
> 
> No, I meant kubernetes validating webhook, like https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation

ok, but then it is a totally different story which will add new webhook in api to validate fields etc.
I started this way  a couple of months ago but did not continue with it due to other high priority tasks, but if you would like you can create a ticket and work on this.

**Comment by ykaliuta** [comment]
> > > > Just thought a bit about that. May be there should exist some sort of Webhook to prevent such configurations at the first place?
> > > 
> > > 
> > > i do not want to enforce git webhook (if this is the webhook you meant) or it will be a new commit everytime and difficult for reviewer, we do have squash-merge for PR so the committer (or reviewer) is responsibile to get a clean history for the PR before merge
> > 
> > 
> > No, I meant kubernetes validating webhook, like https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation
> 
> ok, but then it is a totally different story which will add new webhook in api to validate fields etc.

Sure, it's a further enhancement and it's good to get proper reporting first (I've updated the PR BTW and added phaseError set like in some other places)

> I started this way a couple of months ago but did not continue with it due to other high priority tasks, but if you would like you can create a ticket and work on this.

Sounds great actually!

**Comment by zdtsw** [comment]
create issue : https://github.com/opendatahub-io/opendatahub-operator/issues/693

**Review by zdtsw — APPROVED** [review]
/lgtm

File: controllers/datasciencecluster/datasciencecluster_controller.go
Comment (zdtsw) [inline controllers/datasciencecluster/datasciencecluster_controller.go:145]: why do we have check again on the number of DSCI instance? 
if we have it here, do we still need the block if len(instances.Items) > 1 {...}

Comment (ykaliuta) [reply controllers/datasciencecluster/datasciencecluster_controller.go:145]: In datasciencecluster_controller.go file? `if len(instances.Items) > 1 `  checks for DataScienceCluster objects, but here it is for DSCInitialization one.

**Pull request merged**
